marimo/_save/hash.py

```diff
@@ -4,106 +4,246 @@
-        sha.update(bytes("|".join(code_obj.co_names), "utf8"))
-        sha.update(code_obj.co_code)
+        hash_alg.update(bytes("|".join(code_obj.co_names), "utf8"))
+        hash_alg.update(code_obj.co_code)
 
     process(code)
-    return sha.digest()
+    return hash_alg.digest()
 
 
-def hash_raw_module(module: ast.Module) -> bytes:
+def hash_raw_module(
+    module: ast.Module, hash_type: str = DEFAULT_HASH
+) -> bytes:
     # AST has to be compiled to code object prior to process.
     return hash_module(
         compile(
             module,
             "<hash>",
             mode="exec",
             flags=ast.PyCF_ALLOW_TOP_LEVEL_AWAIT,
-        )
+        ),
+        hash_type,
     )
 
 
-def hash_cell_impl(cell: CellImpl) -> bytes:
-    return hash_module(cell.body) + hash_module(cell.last_expr)
+def hash_cell_impl(cell: CellImpl, hash_type: str = DEFAULT_HASH) -> bytes:
+    return hash_module(cell.body, hash_type) + hash_module(
+        cell.last_expr, hash_type
+    )
 
 
-def build_execution_hash(
-    graph: DirectedGraph, cell_id: CellId_t
-) -> Optional[ValidCacheSha]:
+def hash_and_dequeue_execution_refs(
+    hash_alg: Hash,
+    cell_id: CellId_t,
+    graph: DirectedGraph,
+    refs: set[str],
+) -> None:
     # Execution path works by just analyzing the input cells to hash.
-    sha = hashlib.sha256()
     ancestors = graph.ancestors(cell_id)
-    references = sorted(
-        [hash_cell_impl(graph.cells[cell_id]) for cell_id in ancestors]
+    # Prune to only the ancestors that are tied to the references.
+    ref_cells = set().union(
+        *[graph.definitions.get(ref, set()) for ref in refs]
     )
-    for code in references:
-        sha.update(code)
-    return ValidCacheSha(sha, "ExecutionPath")
+    to_hash = ancestors & ref_cells
+    for ancestor_id in sorted(to_hash):
+        cell_impl = graph.cells[ancestor_id]
+        hash_alg.update(hash_cell_impl(cell_impl, hash_alg.name))
+        for ref in cell_impl.defs:
+            # Look for both, since mangle reference depends on the context of
+            # the definition.
+            if ref in refs:
+                refs.remove(ref)
+            unmangled_ref, _ = unmangle_local(ref)
+            if unmangled_ref in refs:
+                refs.remove(unmangled_ref)
 
 
-def build_content_hash(
-    graph: DirectedGraph,
+def standardize_tensor(tensor: Tensor) -> Optional[Tensor]:
+    # TODO: Consider moving to a more general utility module.
+    if hasattr(tensor, "__array__") or hasattr(tensor, "toarray"):
+        if not hasattr(tensor, "__array_interface__"):
+            DependencyManager.numpy.require(
+                "to render images from generic arrays in `mo.image`"
+            )
+            import numpy
+
+            # Capture those sparse cases
+            if hasattr(tensor, "toarray"):
+                tensor = tensor.toarray()
+            tensor = numpy.array(tensor)
```

Comment:
For tensors on GPU, do you need to move them to CPU? I think torch raises when trying to convert to numpy without first moving to CPU. Just checking ...